core/types/bal.go

```diff
@@ -0,0 +1,227 @@
+func (b balanceDiff) copy() balanceDiff {
+	res := make(balanceDiff)
+	for idx, balance := range b {
+		res[idx] = balance.Clone()
+	}
+	return res
+}
+
+// PrettyPrint returns a human-readable representation of the access list
+func (b *BlockAccessList) PrettyPrint() string {
+	enc := b.toEncodingObj()
+	return enc.prettyPrint()
+}
+
+// Hash computes the SSZ hash of the access list
+func (b *BlockAccessList) Hash() common.Hash {
+	hash, err := b.toEncodingObj().HashTreeRoot()
+	if err != nil {
+		// errors here are related to BAL values exceeding maximum size defined
+		// by the spec.  Hard-fail because these cases are not expected to be hit
+		// under reasonable conditions.
+		panic(err)
+	}
+	return hash
+}
+
+// codeChange contains the code deployed at an address and the transaction
+// index where the deployment took place.
+type codeChange struct {
+	TxIndex uint16
+	Code    []byte `json:"code,omitempty"`
+}
+
+// post-state values of an account's storage slots modified in a block, keyed
+// by slot key
+type storageWrites map[common.Hash]slotWrites
+
+func (s storageWrites) copy() storageWrites {
+	res := make(storageWrites)
+	for slot, writes := range s {
+		res[slot] = maps.Clone(writes)
+	}
+	return res
+}
+
+// accountAccess contains post-block account state for mutations as well as
+// all storage keys that were read during execution.
+type accountAccess struct {
+	StorageWrites  storageWrites            `json:"storageWrites,omitempty"`
+	StorageReads   map[common.Hash]struct{} `json:"storageReads,omitempty"`
+	BalanceChanges balanceDiff              `json:"balanceChanges,omitempty"`
+	NonceChanges   accountNonceDiffs        `json:"nonceChanges,omitempty"`
+
+	// only set for contract accounts which were deployed in the block
+	CodeChange *codeChange `json:"codeChange,omitempty"`
+}
+
+func newAccountAccess() *accountAccess {
+	return &accountAccess{
+		StorageWrites:  make(map[common.Hash]slotWrites),
+		StorageReads:   make(map[common.Hash]struct{}),
+		BalanceChanges: make(balanceDiff),
+		NonceChanges:   make(accountNonceDiffs),
+	}
+}
+
+// the post-state nonce values of a contract account keyed by tx index
+type accountNonceDiffs map[uint16]uint64
+
+// the post-state values of a storage slot, keyed by tx index
+type slotWrites map[uint16]common.Hash
+
+// Copy returns a deep copy of the access list.
+func (b *BlockAccessList) Copy() *BlockAccessList {
+	res := new(BlockAccessList)
+	for addr, aa := range b.accounts {
+		var aaCopy accountAccess
+		aaCopy.StorageWrites = aa.StorageWrites.copy()
+		aaCopy.StorageReads = maps.Clone(aa.StorageReads)
+		aaCopy.BalanceChanges = aa.BalanceChanges.copy()
+		aaCopy.NonceChanges = maps.Clone(aa.NonceChanges)
+		if aa.CodeChange != nil {
+			aaCopy.CodeChange = &codeChange{
+				TxIndex: aa.CodeChange.TxIndex,
+				Code:    bytes.Clone(aa.CodeChange.Code),
+			}
+		}
+		res.accounts[addr] = &aaCopy
+	}
+	return res
+}
+
+func (e *encodingBlockAccessList) prettyPrint() string {
```

Comment:
this should probably be moved? or not be the `encodingBlockAccessList` type

Reply:
It's called by `BlockAccess.PrettyPrint`:  I convert to the encoding format before printing so that the output of `PrettyPrint` is consistent, instead of random iteration order over maps in the `BlockAccessList`.